recipes-extended/quark/files/add-POST-and-v4v-support.patch

```diff
@@ -0,0 +1,319 @@
+index e9d0a06..a3eebf6 100644
+--- a/main.c
++++ b/main.c
+@@ -13,6 +13,7 @@
+ #include <stdio.h>
+ #include <stdlib.h>
+ #include <string.h>
++#include <syslog.h>
+ #include <time.h>
+ #include <unistd.h>
+ 
+@@ -31,14 +32,17 @@ serve(int infd, struct sockaddr_storage *in_sa)
+ 	char inaddr[INET6_ADDRSTRLEN /* > INET_ADDRSTRLEN */];
+ 	char tstmp[21];
+ 
+-	/* set connection timeout */
+-	if (sock_set_timeout(infd, 30)) {
+-		goto cleanup;
+-	}
+-
+ 	/* handle request */
+-	if (!(status = http_get_request(infd, &r))) {
+-		status = http_send_response(infd, &r);
++	status = http_get_request(infd, &r);
++
++	if (!status) {
++		if (r.method == M_POST) {
++			/* handle POST */
++			syslog(LOG_INFO, "[quark] %s", r.payload);
++		} else {
++			/* handle GET */
++			status = http_send_response(infd, &r);
++		}
+ 	}
+ 
+ 	/* write output to log */
+@@ -228,18 +232,6 @@ main(int argc, char *argv[])
+ 		die("setrlimit RLIMIT_NPROC:");
+ 	}
+ 
+-	/* validate user and group */
+-	errno = 0;
+-	if (user && !(pwd = getpwnam(user))) {
+-		die("getpwnam '%s': %s", user, errno ? strerror(errno) :
+-		    "Entry not found");
+-	}
+-	errno = 0;
+-	if (group && !(grp = getgrnam(group))) {
+-		die("getgrnam '%s': %s", group, errno ? strerror(errno) :
+-		    "Entry not found");
+-	}
+-
+ 	/* Open a new process group */
+ 	setpgid(0,0);
+ 
+@@ -270,23 +262,6 @@ main(int argc, char *argv[])
+ 			die("chroot .:");
+ 		}
+ 
+-		/* drop root */
+-		if (grp && setgroups(1, &(grp->gr_gid)) < 0) {
+-			die("setgroups:");
+-		}
+-		if (grp && setgid(grp->gr_gid) < 0) {
+-			die("setgid:");
+-		}
+-		if (pwd && setuid(pwd->pw_uid) < 0) {
+-			die("setuid:");
+-		}
+-		if (getuid() == 0) {
+-			die("Won't run as root user", argv0);
+-		}
+-		if (getgid() == 0) {
+-			die("Won't run as root group", argv0);
+-		}
+-
+ 		/* accept incoming connections */
+ 		while (1) {
+ 			in_sa_len = sizeof(in_sa);
+diff --git a/resp.c b/resp.c
+index 3075c28..ee12606 100644
+--- a/resp.c
++++ b/resp.c
+@@ -141,6 +141,7 @@ resp_file(int fd, char *name, struct request *r, struct stat *st, char *mime,
+ 	            "Date: %s\r\n"
+ 	            "Connection: close\r\n"
+ 	            "Last-Modified: %s\r\n"
++	            "Expires: Wed, 1 Jan 2030 00:00:00 GMT\r\n"
+ 	            "Content-Type: %s\r\n"
+ 	            "Content-Length: %zu\r\n",
+ 	            s, status_str[s], timestamp(time(NULL), t1),
+diff --git a/util.c b/util.c
+index 0b05d91..5e16271 100644
+--- a/util.c
++++ b/util.c
+@@ -133,3 +133,27 @@ reallocarray(void *optr, size_t nmemb, size_t size)
+ 	}
+ 	return realloc(optr, size * nmemb);
+ }
++
++void
++wait_fd_ready(int fd, bool read, bool write)
++{
++	struct timeval tv;
++	fd_set readfds;
++	fd_set writefds;
++
++	tv.tv_sec = 0;
++	tv.tv_usec = 500000;
++
++	FD_ZERO(&readfds);
++	FD_ZERO(&writefds);
++
++	if (read) {
++		FD_SET(fd, &readfds);
++	}
++
++	if (write) {
++		FD_SET(fd, &writefds);
++	}
++
++	select(fd + 1, &readfds, &writefds, NULL, &tv);
```

Comment:
Concur. Also, since you have a timeout specified and the function returns void, you don't know which occurred when calling this function, either read access, timeout, or interrupt. The return of the select() can help make this determination clearer, or allow you to put in a bit of a spinlock if you want to keep this function returning void.